modules/rostests/apitests/kernel32/UEFIFirmware.c

```diff
@@ -0,0 +1,253 @@
+                                               _A2W(EFI_DUMMY_NAMESPACE_GUID_STRING),
+                                               NULL,
+                                               0);
+    dwError = GetLastError();
+    ok(dwLength == 0, "dwLength = %lu, expected 0\n", dwLength);
+
+    bIsUEFI = (dwLength == 0 && dwError != ERROR_INVALID_FUNCTION);
+    test_GetFirmwareType(bIsUEFI);
+    if (!bIsUEFI)
+    {
+        skip("Skipping tests that require UEFI environment.\n");
+        return;
+    }
+
+    /* Test ANSI function too */
+    dwLengthTemp = GetFirmwareEnvironmentVariableA(EFI_DUMMY_VARIABLE_NAME,
+                                                   EFI_DUMMY_NAMESPACE_GUID_STRING,
+                                                   NULL,
+                                                   0);
+    dwErrorTemp = GetLastError();
+    ok(dwLengthTemp == dwLength && dwErrorTemp == dwError,
+       "dwLength = %lu, LastError = %lu, expected bResult = %lu, LastError = %lu\n",
+       dwLengthTemp,
+       dwErrorTemp,
+       dwLength,
+       dwError);
+
+    /* Generate a random variable value to be used in this test */
+    RandomSeed = GetTickCount();
+    EfiVariableValue = RtlRandom(&RandomSeed);
+
+    /* Try to set firmware variable, should fail with ERROR_PRIVILEGE_NOT_HELD,
+     * because no SE_SYSTEM_ENVIRONMENT_NAME privilege enabled by default. */
+    bResult = SetFirmwareEnvironmentVariableW(_A2W(EFI_TEST_VARIABLE_NAME),
+                                              _A2W(EFI_TEST_GUID_STRING),
+                                              &EfiVariableValue,
+                                              sizeof(EfiVariableValue));
+    dwError = GetLastError();
+    ok(!bResult && dwError == ERROR_PRIVILEGE_NOT_HELD,
+       "bResult = %d, LastError = %lu, expected bResult = 0, LastError = ERROR_PRIVILEGE_NOT_HELD\n",
+       bResult,
+       dwError);
+
+    /* Test ANSI function too */
+    bResultTemp = SetFirmwareEnvironmentVariableA(EFI_TEST_VARIABLE_NAME,
+                                                  EFI_TEST_GUID_STRING,
+                                                  &EfiVariableValue,
+                                                  sizeof(EfiVariableValue));
+    dwErrorTemp = GetLastError();
+    ok(bResultTemp == bResult && dwErrorTemp == dwError,
+       "bResult = %d, LastError = %lu, expected bResult = %d, LastError = %lu\n",
+       bResultTemp,
+       dwErrorTemp,
+       bResult,
+       dwError);
+
+    /* Enable SE_SYSTEM_ENVIRONMENT_NAME privilege required by the following tests */
+    bResult = OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken);
+    if (!bResult)
+    {
+        skip("OpenProcessToken failed with error: 0x%08lX, aborting.\n", GetLastError());
+        return;
+    }
+    Privilege.PrivilegeCount = 1;
+    Privilege.Privileges[0].Luid = (LUID){ SE_SYSTEM_ENVIRONMENT_PRIVILEGE, 0 };
```

Comment:
Is the cast needed?

Note: otherwise, `Privilege.Privileges[0].Luid = RtlConvertLongToLuid(SE_SYSTEM_ENVIRONMENT_PRIVILEGE);`

Reply:
Yes, the cast is needed, because here is not a variable declaration like `LUID Luid = { SE_SYSTEM_ENVIRONMENT_PRIVILEGE, 0 }`.
`RtlConvertLongToLuid` is better but seems it exists in XDK/DDK, for KM only.

Reply:
Please avoid ugly casts like these as there is no reason for it, thanks.

You could simply use `RtlConvertUlongToLuid` as it is defined in NDK and since you are writing an apitest in Kernel32 it already imports the RTL functions from here -> https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/apitests/kernel32/precomp.h;hb=961893a712a0ba3a508b743aac2a769500397d0d#l16

Reply:
Committed.